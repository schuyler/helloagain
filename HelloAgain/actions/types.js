'use strict'

// Contacts
export const TOGGLE_ACTIVE = 'TOGGLE_ACTIVE'

// Friends
export const UPDATE_FRIEND = 'UPDATE_FRIEND'
