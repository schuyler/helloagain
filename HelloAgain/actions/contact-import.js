'use strict'

import Contacts from 'react-native-contacts'

import {
  CONTACTS_LOADED, CONTACT_LOAD_FAILED
} from './types';

export const contactsLoaded = (contacts) => {
  return {type: CONTACTS_LOADED, contacts: contacts}
}

export const contactLoadFailed = (error) => {
  return {type: CONTACT_LOAD_FAILED, error: error}
}

export const loadNativeContacts = () => {
  // Create a Promise that gets all native contacts and resolves on success.
  const loadContacts = new Promise((resolve, reject) => {
    Contacts.getAll((err, contacts) => {
      err ? reject(err) : resolve(contacts)
    })
  })
  // Return an action thunk that:
  //   1. Loads all contacts
  //   2. On success, dispatches the `contactsLoaded` action
  //   3. Then, dispatches the `writeIDsForAllFriends` to write back any new
  //      IDs to the native contact store
  //   4. On failure, dispatch the `contactLoadFailed` action
  return dispatch => {
    return loadContacts
      .then(contacts => {
        let result = dispatch(contactsLoaded(contacts))
        console.log("contactsLoaded:", Object.values(result).length)
      })
      .then(_ => { return dispatch(writeIDsForAllFriends()) })
      .catch(error => dispatch(contactLoadFailed(error)))
  }
}

const generateURL = (item) => {
  return "helloagain://to/" + item.helloAgainID
}

// Check a single contact record to see if it needs the HelloAgain URL
// written to it, and if it does, write it.
export const writeIDtoNativeContact = (item) => {
  // Never been active? Don't touch the native contact.
  if (item.isActive === undefined) {
    return
  }
  // Are any of the contact's urlAddresses a HelloAgain address? Great. Skip
  // this one.
  let matchingURLs = (item.urlAddresses || []).filter(url => url.label == "HelloAgain")
  if (matchingURLs.length > 0) {
    return
  }
  // Otherwise, generate the URL, push it onto the urlAddresses list with the
  // right label, and then update the native contact store, with a log message
  // either way.
  let url = generateURL(item)
  item.urlAddresses.push({label: "HelloAgain", urlAddress: url})
  Contacts.updateContact(item, (err, contact) => {
    if (err) {
      console.log(`Error updating ${item.givenName} ${item.familyName} ` +
                  `with URL ${url}:`, err, contact)
    } else {
      console.log(`Wrote URL to Contacts: ${url}`)
    }
  })
}

const writeIDsForAllFriends = () => {
  // Return an action thunk that is pure side-effects -- just iterate
  // over the entire set of contacts and make sure that any once-active friends
  // that don't have HelloAgain URLs associated with them get one. This
  // shouldn't really ever happen, because the `toggleActive` action takes care
  // of the records one-by-one, but I guess it can't hurt to be certain.
  return (_, getState) => {
    const state = getState()
    const friends = Object.values(state.friends).filter(f => f.isActive !== undefined)
    friends.forEach(item => writeIDtoNativeContact(item))
  }
}
