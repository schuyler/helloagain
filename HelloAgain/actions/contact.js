'use strict'

import { TOGGLE_ACTIVE } from './types';

export const toggleActive = (friend) => {
  return { type: TOGGLE_ACTIVE, friend: friend }
}
