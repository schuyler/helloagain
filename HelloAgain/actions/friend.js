'use strict'

import {
  UPDATE_FRIEND,
  MARK_AS_CONTACTED
} from './types'

export const updateFriend = (friend) => {
  return { type: UPDATE_FRIEND, friend: friend }
}

export const markAsContacted = (friend) => {
  return { type: MARK_AS_CONTACTED, friend: friend }
}
