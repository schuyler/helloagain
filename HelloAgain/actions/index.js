'use strict';

// Enumerate the dispatchable actions.
export const UPDATE_FRIEND = 'UPDATE_FRIEND'

// Helpers for dispatching actions.
export const updateFriend = (friend) => {
  return { type: UPDATE_FRIEND, friend: friend }
}
