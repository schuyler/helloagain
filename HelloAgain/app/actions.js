export const IMPORT_CONTACTS = 'IMPORT_CONTACTS';

export const UPDATE_FRIEND = 'UPDATE_FRIEND';

export const FRIEND_CONTACTED = 'FRIEND_CONTACTED';

export const UPDATE_SETTINGS = 'UPDATE_SETTINGS';

export function importContacts(contactList) {
  return { type: IMPORT_CONTACTS, contactList };
}

export function updateFriend(friend) {
  return { type: UPDATE_FRIEND, friend };
}
