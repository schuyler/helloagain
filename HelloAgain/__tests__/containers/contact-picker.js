'use strict'

import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme'
import configureStore from 'redux-mock-store'

import { TOGGLE_ACTIVE } from "../../actions/types"
import ContactPicker from "../../containers/contact-picker"
import FRIENDS_FIXTURE from "../fixtures/friends"

const mockStore = configureStore([])
const store = mockStore({friends: FRIENDS_FIXTURE})

it('renders a list of contacts', () => {
  expect(renderer.create(
    <ContactPicker store={store} />
  )).toMatchSnapshot()
})

it('picks up a sorted list of contacts from the store', () => {
  const wrapper = shallow(
    <ContactPicker store={store} />
  )
  const items = wrapper.prop("items")
  expect(items).toMatchSnapshot()
})

it('can toggle a contact', () => {
  const wrapper = shallow(
    <ContactPicker store={store} />
  )
  const items = wrapper.prop("items")
  const contact = items[0]
  wrapper.props().onContactPress(contact)
  const actions = store.getActions()
  expect(actions[0]).toMatchObject({ type: TOGGLE_ACTIVE, friend: contact })
})
