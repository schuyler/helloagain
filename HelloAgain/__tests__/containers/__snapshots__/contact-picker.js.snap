exports[`test picks up a sorted list of contacts from the store 1`] = `
Array [
  Object {
    "familyName": "Messier",
    "givenName": "Charles",
    "recordID": 6,
  },
  Object {
    "familyName": "Tombaugh",
    "givenName": "Clyde",
    "recordID": 3,
  },
  Object {
    "familyName": "Klumpke",
    "givenName": "Dorothea",
    "recordID": 15,
  },
  Object {
    "familyName": "Hubble",
    "givenName": "Edwin",
    "recordID": 10,
  },
  Object {
    "familyName": "Herzsprung",
    "givenName": "Ejnar",
    "recordID": 9,
  },
  Object {
    "familyName": "Struve",
    "givenName": "Gustav Wilhelm Ludvig",
    "recordID": 16,
  },
  Object {
    "familyName": "Leavitt",
    "givenName": "Henrietta",
    "middleName": "Swan",
    "recordID": 7,
  },
  Object {
    "familyName": "Oort",
    "givenName": "Jan",
    "recordID": 14,
  },
  Object {
    "familyName": "Kepler",
    "givenName": "Johannes",
    "recordID": 12,
  },
  Object {
    "familyName": "Herschel",
    "givenName": "John",
    "recordID": 5,
  },
  Object {
    "familyName": "Lalande",
    "givenName": "Jérôme",
    "recordID": 8,
  },
  Object {
    "familyName": "Wolf",
    "givenName": "Max",
    "recordID": 17,
  },
  Object {
    "familyName": "Humason",
    "givenName": "Milton",
    "middleName": "L.",
    "recordID": 13,
  },
  Object {
    "familyName": "Copernicus",
    "givenName": "Nicolaus",
    "recordID": 1,
  },
  Object {
    "familyName": "Brahe",
    "givenName": "Tycho",
    "recordID": 2,
  },
  Object {
    "familyName": "Rubin",
    "givenName": "Vera",
    "recordID": 11,
  },
  Object {
    "familyName": "Herschel",
    "givenName": "William",
    "recordID": 4,
  },
]
`;

exports[`test renders a list of contacts 1`] = `
<ScrollView
  dataSource={
    ListViewDataSource {
      "items": 17,
    }
  }
  initialListSize={10}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={1000}
  onKeyboardDidHide={undefined}
  onKeyboardDidShow={undefined}
  onKeyboardWillHide={undefined}
  onKeyboardWillShow={undefined}
  onLayout={[Function]}
  onScroll={[Function]}
  pageSize={1}
  removeClippedSubviews={true}
  renderRow={[Function]}
  scrollEventThrottle={50}
  scrollRenderAheadDistance={1000}
  stickyHeaderIndices={Array []}
  style={
    Object {
      "alignSelf": "stretch",
      "backgroundColor": "#ffffff",
      "paddingTop": 20,
    }
  }>
  <View
    accessibilityComponentType={undefined}
    accessibilityLabel={undefined}
    accessibilityTraits={undefined}
    accessible={true}
    hitSlop={undefined}
    onLayout={undefined}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Array [
        Object {
          "backgroundColor": "transparent",
        },
        undefined,
      ]
    }
    testID={undefined}>
    <View
      style={
        Object {
          "backgroundColor": "#F5FCFF",
          "flex": 1,
          "flexDirection": "row",
          "justifyContent": "space-around",
          "padding": 5,
        }
      }>
      <Image
        source={Object {}}
        style={undefined} />
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "flex": 5,
            "fontSize": 20,
            "marginBottom": 5,
            "textAlign": "left",
          }
        }>
        Charles
         
        Messier
      </Text>
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "fontSize": 25,
            "marginRight": 15,
            "textAlign": "right",
          }
        }>
        
      </Text>
    </View>
  </View>
  <View
    accessibilityComponentType={undefined}
    accessibilityLabel={undefined}
    accessibilityTraits={undefined}
    accessible={true}
    hitSlop={undefined}
    onLayout={undefined}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Array [
        Object {
          "backgroundColor": "transparent",
        },
        undefined,
      ]
    }
    testID={undefined}>
    <View
      style={
        Object {
          "backgroundColor": "#F5FCFF",
          "flex": 1,
          "flexDirection": "row",
          "justifyContent": "space-around",
          "padding": 5,
        }
      }>
      <Image
        source={Object {}}
        style={undefined} />
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "flex": 5,
            "fontSize": 20,
            "marginBottom": 5,
            "textAlign": "left",
          }
        }>
        Clyde
         
        Tombaugh
      </Text>
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "fontSize": 25,
            "marginRight": 15,
            "textAlign": "right",
          }
        }>
        
      </Text>
    </View>
  </View>
  <View
    accessibilityComponentType={undefined}
    accessibilityLabel={undefined}
    accessibilityTraits={undefined}
    accessible={true}
    hitSlop={undefined}
    onLayout={undefined}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Array [
        Object {
          "backgroundColor": "transparent",
        },
        undefined,
      ]
    }
    testID={undefined}>
    <View
      style={
        Object {
          "backgroundColor": "#F5FCFF",
          "flex": 1,
          "flexDirection": "row",
          "justifyContent": "space-around",
          "padding": 5,
        }
      }>
      <Image
        source={Object {}}
        style={undefined} />
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "flex": 5,
            "fontSize": 20,
            "marginBottom": 5,
            "textAlign": "left",
          }
        }>
        Dorothea
         
        Klumpke
      </Text>
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "fontSize": 25,
            "marginRight": 15,
            "textAlign": "right",
          }
        }>
        
      </Text>
    </View>
  </View>
  <View
    accessibilityComponentType={undefined}
    accessibilityLabel={undefined}
    accessibilityTraits={undefined}
    accessible={true}
    hitSlop={undefined}
    onLayout={undefined}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Array [
        Object {
          "backgroundColor": "transparent",
        },
        undefined,
      ]
    }
    testID={undefined}>
    <View
      style={
        Object {
          "backgroundColor": "#F5FCFF",
          "flex": 1,
          "flexDirection": "row",
          "justifyContent": "space-around",
          "padding": 5,
        }
      }>
      <Image
        source={Object {}}
        style={undefined} />
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "flex": 5,
            "fontSize": 20,
            "marginBottom": 5,
            "textAlign": "left",
          }
        }>
        Edwin
         
        Hubble
      </Text>
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "fontSize": 25,
            "marginRight": 15,
            "textAlign": "right",
          }
        }>
        
      </Text>
    </View>
  </View>
  <View
    accessibilityComponentType={undefined}
    accessibilityLabel={undefined}
    accessibilityTraits={undefined}
    accessible={true}
    hitSlop={undefined}
    onLayout={undefined}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Array [
        Object {
          "backgroundColor": "transparent",
        },
        undefined,
      ]
    }
    testID={undefined}>
    <View
      style={
        Object {
          "backgroundColor": "#F5FCFF",
          "flex": 1,
          "flexDirection": "row",
          "justifyContent": "space-around",
          "padding": 5,
        }
      }>
      <Image
        source={Object {}}
        style={undefined} />
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "flex": 5,
            "fontSize": 20,
            "marginBottom": 5,
            "textAlign": "left",
          }
        }>
        Ejnar
         
        Herzsprung
      </Text>
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "fontSize": 25,
            "marginRight": 15,
            "textAlign": "right",
          }
        }>
        
      </Text>
    </View>
  </View>
  <View
    accessibilityComponentType={undefined}
    accessibilityLabel={undefined}
    accessibilityTraits={undefined}
    accessible={true}
    hitSlop={undefined}
    onLayout={undefined}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Array [
        Object {
          "backgroundColor": "transparent",
        },
        undefined,
      ]
    }
    testID={undefined}>
    <View
      style={
        Object {
          "backgroundColor": "#F5FCFF",
          "flex": 1,
          "flexDirection": "row",
          "justifyContent": "space-around",
          "padding": 5,
        }
      }>
      <Image
        source={Object {}}
        style={undefined} />
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "flex": 5,
            "fontSize": 20,
            "marginBottom": 5,
            "textAlign": "left",
          }
        }>
        Gustav Wilhelm Ludvig
         
        Struve
      </Text>
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "fontSize": 25,
            "marginRight": 15,
            "textAlign": "right",
          }
        }>
        
      </Text>
    </View>
  </View>
  <View
    accessibilityComponentType={undefined}
    accessibilityLabel={undefined}
    accessibilityTraits={undefined}
    accessible={true}
    hitSlop={undefined}
    onLayout={undefined}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Array [
        Object {
          "backgroundColor": "transparent",
        },
        undefined,
      ]
    }
    testID={undefined}>
    <View
      style={
        Object {
          "backgroundColor": "#F5FCFF",
          "flex": 1,
          "flexDirection": "row",
          "justifyContent": "space-around",
          "padding": 5,
        }
      }>
      <Image
        source={Object {}}
        style={undefined} />
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "flex": 5,
            "fontSize": 20,
            "marginBottom": 5,
            "textAlign": "left",
          }
        }>
        Henrietta
         
        Leavitt
      </Text>
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "fontSize": 25,
            "marginRight": 15,
            "textAlign": "right",
          }
        }>
        
      </Text>
    </View>
  </View>
  <View
    accessibilityComponentType={undefined}
    accessibilityLabel={undefined}
    accessibilityTraits={undefined}
    accessible={true}
    hitSlop={undefined}
    onLayout={undefined}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Array [
        Object {
          "backgroundColor": "transparent",
        },
        undefined,
      ]
    }
    testID={undefined}>
    <View
      style={
        Object {
          "backgroundColor": "#F5FCFF",
          "flex": 1,
          "flexDirection": "row",
          "justifyContent": "space-around",
          "padding": 5,
        }
      }>
      <Image
        source={Object {}}
        style={undefined} />
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "flex": 5,
            "fontSize": 20,
            "marginBottom": 5,
            "textAlign": "left",
          }
        }>
        Jan
         
        Oort
      </Text>
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "fontSize": 25,
            "marginRight": 15,
            "textAlign": "right",
          }
        }>
        
      </Text>
    </View>
  </View>
  <View
    accessibilityComponentType={undefined}
    accessibilityLabel={undefined}
    accessibilityTraits={undefined}
    accessible={true}
    hitSlop={undefined}
    onLayout={undefined}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Array [
        Object {
          "backgroundColor": "transparent",
        },
        undefined,
      ]
    }
    testID={undefined}>
    <View
      style={
        Object {
          "backgroundColor": "#F5FCFF",
          "flex": 1,
          "flexDirection": "row",
          "justifyContent": "space-around",
          "padding": 5,
        }
      }>
      <Image
        source={Object {}}
        style={undefined} />
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "flex": 5,
            "fontSize": 20,
            "marginBottom": 5,
            "textAlign": "left",
          }
        }>
        Johannes
         
        Kepler
      </Text>
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "fontSize": 25,
            "marginRight": 15,
            "textAlign": "right",
          }
        }>
        
      </Text>
    </View>
  </View>
  <View
    accessibilityComponentType={undefined}
    accessibilityLabel={undefined}
    accessibilityTraits={undefined}
    accessible={true}
    hitSlop={undefined}
    onLayout={undefined}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Array [
        Object {
          "backgroundColor": "transparent",
        },
        undefined,
      ]
    }
    testID={undefined}>
    <View
      style={
        Object {
          "backgroundColor": "#F5FCFF",
          "flex": 1,
          "flexDirection": "row",
          "justifyContent": "space-around",
          "padding": 5,
        }
      }>
      <Image
        source={Object {}}
        style={undefined} />
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "flex": 5,
            "fontSize": 20,
            "marginBottom": 5,
            "textAlign": "left",
          }
        }>
        John
         
        Herschel
      </Text>
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "fontSize": 25,
            "marginRight": 15,
            "textAlign": "right",
          }
        }>
        
      </Text>
    </View>
  </View>
</ScrollView>
`;
